librbd: make mirror group enable asynchronous

ajarr opened the pull request:
Notable enhancements have been made to the mirror group_enable API:
* Previously, member images were enabled for mirroring synchronously
  and one after another. They are now enabled asynchronously and
  concurrently, making  the time to enable a group independent of the
  number of member images it contains.

* Intermediate steps involved in enabling an image, such as fetching
  mirror peers, acquiring an exclusive lock, creating a mirror snapshot,
  among other operations, run asynchronously but are synchronized
  across all the group member images using C_Gather callbacks. An
  asynchronous operation on an image proceeds to the next step only
  after that operation completes for all member images. This allows for
  easier cleanup of member images and the group when an intermediate
  operation fails for one or more  member images.

* The exclusive locks on member images are now held for a shorter
  duration. All member image locks are acquired just before taking
  image snapshots and released after retrieving the snapshot IDs.
  Previously, the locks were obtained earlier in the sequence of
  operations, even for steps that did not need them, and were not
  explicitly released.
  
  Doc comparing call flow of the old sync group_enable with the new async group_enable,
  https://docs.google.com/document/d/1Q5sIZlK5rGck7-qkLkFJ515cGAOSiyNKSMb8dGgqqH0/

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

pkalever commented on src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc:
```diff
@@ -0,0 +1,395 @@
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to retrieve mirror peers: " << cpp_strerror(r)
+               << dendl;
```
indentation

pkalever commented on src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc:
```diff
@@ -0,0 +1,395 @@
+
+  m_cct = m_image_ctxs[0]->cct;
+
+  //FIXME: using the first image_ctx pool for now
```
We are already rewriting CreatePrimaryRequest.cc to fit to group mirroring requirements. Many routines needs a rewrite including the way we get mirror peers for different images when we want cross pool images in the group (its again a rewrite). Rather I would suggest keep it generic enough, by using different image pool and their respective peers.

ajarr replied:
Fixed

pkalever commented on src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc:
```diff
@@ -0,0 +1,395 @@
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to create mirror snapshot: " << cpp_strerror(r)
+               << dendl;
```
indentation

ajarr commented on src/librbd/mirror/GroupEnableRequest.cc:
```diff
@@ -0,0 +1,890 @@
+                   << dendl;
+      r = -EINVAL;
+    } else if (
+	m_mirror_group.state == cls::rbd::MIRROR_GROUP_STATE_DISABLING) {
```
We want the mirroring on group to be disabled. This is the case when `r = -ENOENT`. Only if mirroring disabled do we proceed to the next step.

ajarr commented on src/librbd/mirror/GroupEnableRequest.cc:
```diff
@@ -0,0 +1,890 @@
+  if (r < 0) {
+    lderr(m_cct) << "failed to remove mirror group snapshot: " << cpp_strerror(r)
+                 << dendl;
+    close_images();
```
If the undo steps fail, then I expect `mirror group disable --force` to clean it up. Re-issuing the `mirror group enable` won't clean up issues in the undo codepath. This is how the current `group_enable()` API works as per @VinayBhaskar-V . 

@VinayBhaskar-V can you provide your testing results here?

VinayBhaskar-V replied:
Yes, when an undo fails the expected behavior is to run `mirror group disable --force`
I tried with below diff

```
--- a/src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc
+++ b/src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc
@@ -241,10 +241,13 @@ void GroupImageCreatePrimaryRequest<I>::create_snapshots() {
       m_cct, get_default_snap_create_flags(m_image_ctxs[i]),
       &snap_create_flags);
     ceph_assert(r == 0);
-
-    m_image_ctxs[i]->operations->snap_create(ns, m_snap_names[i],
+    if (i != m_image_ctxs.size()-1 ){
+      m_image_ctxs[i]->operations->snap_create(ns, m_snap_names[i],
                                              snap_create_flags, m_prog_ctx,
                                              gather_ctx->new_sub());
+
+    }
+    
   }
 
   gather_ctx->activate();
@@ -253,6 +256,7 @@ void GroupImageCreatePrimaryRequest<I>::create_snapshots() {
 template <typename I>
 void GroupImageCreatePrimaryRequest<I>::handle_create_snapshots(int r) {
   ldout(m_cct, 15) << "r=" << r << dendl;
+   r= -1;
 
   if (r < 0) {
     lderr(m_cct) << "failed to create mirror snapshot: " << cpp_strerror(r)
    
--- a/src/librbd/mirror/snapshot/RemoveGroupSnapshotRequest.cc
+++ b/src/librbd/mirror/snapshot/RemoveGroupSnapshotRequest.cc
@@ -61,7 +61,10 @@ void RemoveGroupSnapshotRequest<I>::remove_group_image_snapshots() {
     ldout(m_cct, 10) << "removing individual snapshot: "
                      << snap.snap_id << ", from image id:"
                      << snap.image_id << dendl;
-    remove_image_snapshot(ictx, snap.snap_id, gather_ctx);
+    if (i != m_group_snap->snaps.size()-1 ){
+      remove_image_snapshot(ictx, snap.snap_id, gather_ctx);
+    }
+    
   }
 
   gather_ctx->activate();
@@ -71,6 +74,7 @@ void RemoveGroupSnapshotRequest<I>::remove_group_image_snapshots() {
 template <typename I>
 void RemoveGroupSnapshotRequest<I>::handle_remove_group_image_snapshots(int r) {
   ldout(m_cct, 10) << "r=" << r << dendl;
+  r = -1;
 
   if (r < 0 && r != -ENOENT) {
     lderr(m_cct) << "failed to remove individual group image snapshots: "

```

```
[root@cephnvme-devel-server-vinaybhaskar build]# bin/rbd --cluster site-a group info data/grp1
rbd group 'grp1':
	id: 1037c803b32f
	mirroring state: disabling
	mirroring mode: snapshot
	mirroring global id: be0ead6b-2544-4743-a22d-cfd9058ce171
	mirroring primary: false
[root@cephnvme-devel-server-vinaybhaskar build]# 
[root@cephnvme-devel-server-vinaybhaskar build]# bin/rbd --cluster site-a group snap ls data/grp1
ID            NAME                                                               STATE       NAMESPACE                                                         
1041473e4eaf  .mirror.primary.be0ead6b-2544-4743-a22d-cfd9058ce171.1041473e4eaf  incomplete  mirror (primary peer_uuids:[491a1615-e985-43b6-b502-90a43e30cfe6])
[root@cephnvme-devel-server-vinaybhaskar build]# 
[root@cephnvme-devel-server-vinaybhaskar build]# bin/rbd --cluster site-b group info data/grp1
rbd: failed to get ID of group grp1: (2) No such file or directory

// recompiled by removing the diff

[root@cephnvme-devel-server-vinaybhaskar build]# 
[root@cephnvme-devel-server-vinaybhaskar build]# bin/rbd --cluster site-a mirror group enable data/grp1
2025-08-08T07:13:18.890+0000 7f84be116640 -1 librbd::mirror::GroupEnableRequest: 0x55c1a25dd600 handle_get_mirror_group: mirroring on group is currently disabling
2025-08-08T07:13:18.891+0000 7f84c5640d40 -1 librbd::api::Mirror: group_enable: failed to mirror enable group: (22) Invalid argument
[root@cephnvme-devel-server-vinaybhaskar build]# 
[root@cephnvme-devel-server-vinaybhaskar build]# bin/rbd --cluster site-a mirror group disable data/grp1
2025-08-08T07:13:37.787+0000 7f2d91447d40 -1 librbd::api::Mirror: group_disable: mirrored group grp1 is not primary, add force option to disable mirroring
[root@cephnvme-devel-server-vinaybhaskar build]# 
[root@cephnvme-devel-server-vinaybhaskar build]# bin/rbd --cluster site-a mirror group disable data/grp1 --force
Mirroring disabled
[root@cephnvme-devel-server-vinaybhaskar build]# 
[root@cephnvme-devel-server-vinaybhaskar build]# bin/rbd --cluster site-a group info data/grp1
rbd group 'grp1':
	id: 1037c803b32f
```

ajarr commented on src/librbd/mirror/GetMirrorImageRequest.cc:
```diff
@@ -0,0 +1,86 @@
+
+  if (r == -ENOENT) {
+    ldout(m_cct, 20) << "mirroring is disabled" << dendl;
+    finish(0);
```
Removed this file. I don't think this comment is relevant anymore.

ajarr commented:
> In this PR do we support adding or removing images from a mirroring group while it is in the `mirror group enabled` state?
> 

No, we don't support dynamic mirror groups.

> If this behavior is supported, would it be possible to add a corresponding test case to validate that image additions/removals during the mirroring-enabled state are handled correctly and that the group snapshot consistency is maintained?

ajarr commented:
The changes can be seen here, https://github.com/ceph/ceph/compare/34e2fb926bff56fa18e098e698726fc76bc38602..6d85553109761fbdebc61897325092e64c20c778

I've added another commit that sets the mirror group to enabling and then creates incomplete group snapshot as done in https://github.com/ceph/ceph/pull/64782/commits/6d85553109761fbdebc61897325092e64c20c778. I need to verify with @nbalacha if this makes sense.

ajarr commented:
Changes to be highlighted in the diff https://github.com/ceph/ceph/compare/34e2fb926bff56fa18e098e698726fc76bc38602..44e05a4f4e2e0f33b852a287339a60297bcffd0d

    Removal of GetMirrorImageRequest state machine; replaced by a lambda callback in GroupEnableRequest state machine
    Swapping steps to create incomplete primary group snapshot and setting mirror group state to enabling in GroupEnableRequest state machine to match the sequence of steps that currently exists in the CGSM PR
    Support for fetching the mirroring peers of each group member image in GroupImageCreatePrimaryRequest state machine
 ,

    Removal of GetMirrorImageRequest state machine; replaced by a lambda callback in GroupEnableRequest state machine
    Swapping steps to create incomplete primary group snapshot and setting mirror group state to enabling in GroupEnableRequest state machine to match the sequence of steps that currently exists in the CGSM PR
    Support for fetching the mirroring peers of each group member image in GroupImageCreatePrimaryRequest state machine

ajarr commented on src/librbd/mirror/GroupEnableRequest.cc:
```diff
@@ -0,0 +1,931 @@
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  m_mirror_images.resize(m_images.size());
+  for (size_t i = 0; i < m_images.size(); i++) {
+    librados::ObjectReadOperation op;
+    cls_client::mirror_image_get_start(&op, m_images[i].spec.image_id);
+
+    bufferlist *out_bl = new bufferlist();
+
+    auto on_mirror_image_get = new LambdaContext(
+      [this, i, out_bl, new_sub_ctx=gather_ctx->new_sub()](int r) {
+        if (r == 0) {
+          auto iter = out_bl->cbegin();
+          r = cls_client::mirror_image_get_finish(&iter, &m_mirror_images[i]);
+        }
+
+        delete out_bl;
+
+        if (r == -ENOENT) {
+          // image is disabled for mirroring as required
+          r = 0;
+        } else if (r == 0) {
+          lderr(m_cct) << "image_id=" << m_images[i].spec.image_id
+                       << " is not disabled for mirroring" << dendl;
+          r = -EINVAL;
+        } else {
+          lderr(m_cct) << "failed to get mirror image info for image_id="
+                       << m_images[i].spec.image_id << dendl;
+        }
+
+        new_sub_ctx->complete(r);
+      });
+
+    auto comp = create_rados_callback(on_mirror_image_get);
+
+    int r = m_group_ioctx.aio_operate(RBD_MIRRORING, comp, &op, out_bl);
+    ceph_assert(r == 0);
+    comp->release();
+  }
```
Your suggestion looks the same as the existing code. What's the change in indentation that is being recommended?

pkalever replied:
Please notice carefully the suggestion is about using auto out_bl = std::make_shared<bufferlist>(); removing delete out_bl; Please copy paste the same to see the differences.

ajarr commented on src/librbd/mirror/GroupEnableRequest.cc:
```diff
@@ -0,0 +1,931 @@
+    &GroupEnableRequest<I>::handle_get_mirror_images_for_cleanup>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  for (size_t i = 0; i < m_images.size(); i++) {
+    librados::ObjectReadOperation op;
+    cls_client::mirror_image_get_start(&op, m_images[i].spec.image_id);
+
+    bufferlist *out_bl = new bufferlist();
+
+    auto on_mirror_image_get = new LambdaContext(
+      [this, i, out_bl, new_sub_ctx=gather_ctx->new_sub()](int r) {
+        if (r == 0) {
+          auto iter = out_bl->cbegin();
+          r = cls_client::mirror_image_get_finish(&iter, &m_mirror_images[i]);
+        }
+
+        delete out_bl;
+
+        if (r == -ENOENT) {
+          r = 0;
+          m_mirror_images[i].state = cls::rbd::MIRROR_IMAGE_STATE_DISABLED;
+        } else if (r < 0) {
+          lderr(m_cct) << "failed to get mirror image info for image_id="
+                       << m_images[i].spec.image_id << dendl;
+        }
+
+        new_sub_ctx->complete(r);
+      });
+
+    auto comp = create_rados_callback(on_mirror_image_get);
+
+    int r = m_group_ioctx.aio_operate(RBD_MIRRORING, comp, &op, out_bl);
+    ceph_assert(r == 0);
+    comp->release();
+  }
```
Your suggestion looks the same as the existing code. What's the change in indentation that is being recommended?

pkalever replied:
Please notice carefully the suggestion is about using `auto out_bl = std::make_shared<bufferlist>(); ` removing `delete out_bl;` Please copy paste the same to see the differences.

ajarr commented on src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc:
```diff
@@ -0,0 +1,421 @@
+
+    ldout(m_cct, 15) << "snap_name=" << m_snap_names[i] << dendl;
+    auto snap_id = m_image_ctxs[i]->get_snap_id(
+      cls::rbd::MirrorSnapshotNamespace{}, m_snap_names[i]);
```
We don't want to overwrite the return code received from refreshing the images. Setting `m_ret_code` to ENOENT here would do that. If creation of a snapshot failed, the `m_ret_code` would already be non-zero.

And I get the point about logging an error if snap_id is CEPH_NOSNAP.  We don't need to check that `m_ret_code==0` here. Irrespective of the value of `m_ret_code`,  if the `snap_id` is zero, then it means that the image snapshot cannot be found

ajarr replied:
How is it possible for `snap_id` to be invalid if `m_ret_code == 0`, which indicates here that creation of image snapshots succeeded, and `r == 0`, which indicates that the refresh of images succeeded?

idryomov commented on src/librbd/mirror/GroupEnableRequest.cc:
```diff
@@ -0,0 +1,931 @@
+      });
+
+    // Open parent as well to check if the image is a clone
+    m_image_ctxs[i]->state->open(OPEN_FLAG_IGNORE_MIGRATING, on_open);
```
> My understanding is that even if one image out of all the group of say 100 images fail to open this entire open_images() will fail in the handle_open_images(), correct?

Yes.  The error that is propagated for the entire `open_images()` would be the error for the first image that failed to open -- this is how `C_Gather` is implemented.

pkalever commented on src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc:
```diff
@@ -0,0 +1,426 @@
+    &GroupImageCreatePrimaryRequest<I>::handle_release_exclusive_locks>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  for (auto ictx: m_image_ctxs) {
+    std::shared_lock owner_lock{ictx->owner_lock};
+    if (ictx->exclusive_lock != nullptr) {
+      ictx->exclusive_lock->release_lock(gather_ctx->new_sub());
+    }
+  }
```
Think it from the per image lock/no-lock situation.

Since we have images in a group here, in our case either all images have locks or all will not have. We might not land in a case where some images might not get lock but some do, at-least I'm not aware of any such case honestly. That said, just to rule them out in all degrees I had the above suggestion as an improvement: 

Just for such kind of situations, the suggestion makesure:
1. complete tracking: all images are properly tracked in the gather context
2. synchronization: gather context will only complete when all sub-contexts are done
3. consistency: prevents operations from continuing before all images are processed

ajarr replied:
As explained in my previous comment, even with the current code, we have 2. and 3 because C_Gather guarantees that.  With your suggestion, I guess we may have additional C_GatherBase log messages for images that don't have exclusive locks. If only some of the member images don't have exclusive locks, maybe better to error out? But we don't error out prior to the code changes here.

@idryomov ^^

idryomov replied:
> As explained in my previous comment, even with the current code, we have 2. and 3 because C_Gather guarantees that.

Right, and I'm failing to see any point in having "1. complete tracking".  Adding a sub-context for an image without exclusive lock feature enabled only to immediately complete that sub-context -- before the parent gather is even activated (!) -- seems pretty pointless to me because it would do nothing other than burning some CPU cycles to grab/release the mutex, increment/decrement the counter for the outstanding sub-contexts and potentially generate a log message in the parent gather.

This

> gather context might complete before all images are processed, potentially causing premature continuation

would be a concern only if the parent gather was activated before all images are iterated through/processed which isn't the case here -- note that `gather_ctx->activate()` call is reached only after `for (auto ictx: m_image_ctxs)` loop completes.

> If only some of the member images don't have exclusive locks, maybe better to error out? But we don't error out prior to the code changes here.

This would need a separate discussion that is out of scope of this PR.  It is indeed the case that if there is no quescing (krbd and the QEMU driver don't support it at all and for rbd-nbd the support is somewhat wonky) and exclusive lock feature isn't enabled on all member images, the group snapshot would be created as usual but may be inconsistent.  However, this is a long-standing behavior going back to the introduction of groups in RBD (initially there was no support for optional quescing so the only thing that could guarantee consistency was exclusive locking and even then exclusive lock feature hasn't been required for member images).

With the above in mind, currently there is no way to "rule them out in all degrees": all respective code has to be prepared for a scenario where some member images have exclusive lock feature enabled and some not.

pkalever commented on src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc:
```diff
@@ -0,0 +1,430 @@
+    ns.group_spec = m_image_ctxs[i]->group_spec;
+    ns.group_snap_id = m_group_snap_id;
+
+    ldout(m_cct, 15) << "name=" << m_snap_names[i] << ", "
+                     << "ns=" << ns << dendl;
```
If you get a chance to respin this again, just consider this minor change:
```suggestion
    ldout(m_cct, 15) << "creating snapshot: image_id=" << m_image_ctxs[i]->id 
                     << ", name=" << m_snap_names[i] << ", "
                     << "ns=" << ns << dendl;
```

pkalever reviewed (approved):
Added a last minute minor log msg suggestion.
Otherwise LGTM.
